pkg/controller/clusters/status.go

```diff
@@ -0,0 +1,187 @@
+	}
+
+	// The workers condition should be marked ready when the total number of nodes specified in the spec is
+	// equal to the ready amount in the cluster and there are no lingering old nodes. We need to check this in
+	// two parts, since in the case of a rolling upgrade, there can be ready machines in the cluster with the old spec.
+
+	totalReady := 0
+	totalUpdatedReplicas := 0
+	totalReplicas := 0
+
+	for _, md := range machineDeployments {
+		// We make sure to check that the status is up to date before using it
+		statusUpToDate := md.Status.ObservedGeneration == md.ObjectMeta.Generation
+		if !statusUpToDate {
+			conditions.MarkFalse(cluster, anywherev1.WorkersReadyConditon, anywherev1.PendingUpdateReason, clusterv1.ConditionSeverityInfo, "Waiting to update machine deployment %s", md.Name)
+			return nil
+		}
+
+		totalReady += int(md.Status.ReadyReplicas)
+		totalUpdatedReplicas += int(md.Status.UpdatedReplicas)
+		totalReplicas += int(md.Status.Replicas)
+	}
+
+	// First, get the number of outdated nodes, as long as there are some, we want to reflect that in the message
+	totalOutdated := totalReplicas - totalUpdatedReplicas
+	if totalOutdated > 0 {
+		conditions.MarkFalse(cluster, anywherev1.WorkersReadyConditon, anywherev1.WaitingForWorkersReadyReason, clusterv1.ConditionSeverityInfo, "Worker nodes not up-to-date yet, %d rolling (%d up to date)", totalReplicas, totalUpdatedReplicas)
+		return nil
+	}
+
+	// Then, we compare the total number of readyReplicas to the expected number,
+	if totalReady != expected {
+		conditions.MarkFalse(cluster, anywherev1.WorkersReadyConditon, anywherev1.WaitingForWorkersReadyReason, clusterv1.ConditionSeverityInfo, "Worker nodes not ready yet, %d expected (%d ready)", expected, totalReady)
+		return nil
+	}
+
+	conditions.MarkTrue(cluster, anywherev1.WorkersReadyConditon)
+	return nil
+}
+
+// UpdateSelfManagedClusterDefaultCNIConfiguredCondition is responsible for updating the DefaultCNIConfiguredCondition for self-managed clusters.
+// This is meant to be done in the CNI recondile, however, self-managed clusters do not use the it, so this status would never get resolved.
+func UpdateSelfManagedClusterDefaultCNIConfiguredCondition(cluster *anywherev1.Cluster) {
+	if !conditions.IsTrue(cluster, anywherev1.ControlPlaneReadyCondition) {
+		return
+	}
+
+	ciliumCfg := cluster.Spec.ClusterNetwork.CNIConfig.Cilium
+
+	// Though it may be installed initially to successfully create the cluster,
+	// if the CNI is configured to skip upgrades, we want to mark the condition as "False"
+	// as maintenance is not being performed by EKS-A.
+
+	// Otherwise, we can assume the CNI has been
+	// configured at this point after confirming the control plane is ready
+	if !ciliumCfg.IsManaged() {
+		conditions.MarkFalse(cluster, anywherev1.DefaultCNIConfiguredCondition, anywherev1.SkipUpgradesForDefaultCNIConfiguredReason, clusterv1.ConditionSeverityWarning, "Maintenance will not be performed during upgrades")
```

Comment:
What does maintenance not being performed mean? Can we just say something like "Skipping upgrade as CNI is not managed by EKS-A"

Reply:
Chris was very against naming the condition ManagedCNI, so I was trying ro keep away from the "managed" verbiage. But yes, we can probably use something more specific